model/src/main/kotlin/utils/Extensions.kt

```diff
@@ -208,3 +212,49 @@ fun List<ScanResult>.filterByProject(project: Project): List<ScanResult> {
         }
     }
 }
+
+/**
+ * Filter [packages][this] to include only [Package]s from [categoriesToInclude]. License choices from [ortResult] are
+ * respected and applied to dismiss licenseCategories that that are not chosen.
+ */
+fun List<Package>.filterByLicenseCategories(
+    categoriesToInclude: Collection<String>,
+    licenseCategorizations: List<LicenseCategorization>,
+    ortResult: OrtResult
+): List<Package> {
+    val packagesByLicense = mutableMapOf<SpdxExpression, MutableSet<Package>>()
+    val licenseInfoResolver = ortResult.createLicenseInfoResolver()
+
+    forEach { pkg ->
+        val resolveLicenseInfo = licenseInfoResolver.resolveLicenseInfo(pkg.id)
+        val effectiveLicenses = resolveLicenseInfo.effectiveLicense(
+            LicenseView.ALL,
+            ortResult.getRepositoryLicenseChoices(),
+            ortResult.getPackageLicenseChoices(pkg.id)
+        )?.decompose().orEmpty()
+
+        effectiveLicenses.forEach { license ->
+            val packages = packagesByLicense[license].orEmpty().toMutableSet()
+            packages += pkg
+
+            packagesByLicense[license] = packages
+        }
+    }
+
+    val categoriesByLicense = mutableMapOf<SpdxExpression, Set<String>>()
+    licenseCategorizations.forEach { licenseCategorization ->
+        val categories = categoriesByLicense[licenseCategorization.id].orEmpty().toMutableSet()
+        categories += licenseCategorization.categories
+
+        categoriesByLicense[licenseCategorization.id] = categories
+    }
+
+    val filteredPackages = mutableListOf<Package>()
+    packagesByLicense.forEach { (license, packages) ->
+        categoriesByLicense[license].orEmpty()
+            .filter { it in categoriesToInclude }
+            .forEach { _ -> filteredPackages += packages }
+    }
```

Comment:
Suggestion:
```
val filteredPackages = packagesByLicense.entries.flatMapTo(mutableSetOf()) { (license, packages) ->
    categoriesByLicense[license].orEmpty().filter { it in categoriesToInclude }
}
```

Reply:
Your suggestion results in a `MutableSet<String>`.
I explicitly chose `List` to not change the data type of `this`.

To get the packages I added `.flatMap { packages }`. Or is there a more elegant way to do this?
```kt
val filteredPackages = packagesByLicense.entries.flatMapTo(mutableListOf()) { (license, packages) ->
    categoriesByLicense[license].orEmpty()
        .filter { it in categoriesToInclude }
        .flatMap { packages }
}
```

Reply:
No longer required due to https://github.com/oss-review-toolkit/ort/pull/3935#discussion_r626502111